mono/metadata/threads.c

```diff
@@ -472,27 +472,113 @@ static void ensure_synch_cs_set (MonoInternalThread *thread)
 	}
 }
 
-static inline void
+static gboolean
 lock_thread (MonoInternalThread *thread)
 {
+	// inc count and read free
+	while (1) {
+		MonoInternalThreadLockFlags flags;
+		flags.i = mono_atomic_load_ptr (&thread->lock_flags.i);
+		if (flags.free)
+			return FALSE;
+		MonoInternalThreadLockFlags old_flags = flags;
+		flags.count += 1;
+		if (mono_atomic_cas_ptr (&thread->lock_flags.i, flags.i, old_flags.i) == old_flags.i)
+			break;
+	}
+
+	// At this point it could be set to free, but we have
+	// a count and we will lock and unlock. Nobody will
+	// free out from under us.
+
 	if (!thread->synch_cs)
 		ensure_synch_cs_set (thread);
 
-	g_assert (thread->synch_cs);
-
 	mono_coop_mutex_lock (thread->synch_cs);
+	return TRUE;
 }
 
-static inline void
+static void
 unlock_thread (MonoInternalThread *thread)
 {
+	// dec count and read free
+	MonoInternalThreadLockFlags flags;
+	flags.i = mono_atomic_load_ptr (&thread->lock_flags.i);
+	while (1) {
+		MonoInternalThreadLockFlags old_flags = flags;
+		g_assertf (flags.count > 0, "count:%d free:%d", flags.count, flags.free);
+		flags.count -= 1;
+		MonoInternalThreadLockFlags old;
+		old.i = mono_atomic_cas_ptr (&thread->lock_flags.i, flags.i, old_flags.i);
+		if (old.i == old_flags.i)
+			break;
+		flags = old;
+	}
+
 	mono_coop_mutex_unlock (thread->synch_cs);
+
+	if (flags.free && flags.count == 0) {
+		MonoCoopMutex *synch_cs = thread->synch_cs;
```

Comment:
Consider cas or swap here, rare path, shouldn't be needed. In fact should be safe to leave the non-null old value.